File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    # 가는 길에 택배 주고 수거까지 해야 효율적
    # 배달이 0이면 cap만큼 수거
    # 수거가 0이면 cap만큼 배달
    # 끝에 계속 확인해야함
    deliveries = deliveries[::-1]
    pickups = pickups[::-1]
    answer = 0
    
    deli = 0
    pick = 0
    for i in range(n):
        deli += deliveries[i]
        pick += pickups[i]
        
        while deli > 0 or pick > 0:
            deli -= cap
            pick -= cap
            answer +=(n-i)*2
    
    return answer

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.js
function solution(cap, n, deliveries, pickups) {
    let answer = 0;
    let dIdx = n - 1;
    let pIdx = n - 1;

    while (dIdx >= 0 || pIdx >= 0) {
       // 가장 먼 거리 찾기
        while (dIdx >= 0 && deliveries[dIdx] === 0) dIdx--;
        while (pIdx >= 0 && pickups[pIdx] === 0) pIdx--;

        if (dIdx < 0 && pIdx < 0) break;

        const dist = Math.max(dIdx, pIdx) + 1;
        answer += dist * 2;

        // cap만큼 배달
        let dCap = cap;
        for (let i = dIdx; i >= 0 && dCap > 0; i--) {
            if (deliveries[i] === 0) continue;
            const deliver = Math.min(deliveries[i], dCap);
            deliveries[i] -= deliver;
            dCap -= deliver;
        }

        // cap만큼 수거
        let pCap = cap;
        for (let i = pIdx; i >= 0 && pCap > 0; i--) {
            if (pickups[i] === 0) continue;
            const pickup = Math.min(pickups[i], pCap);
            pickups[i] -= pickup;
            pCap -= pickup;
        }
    }

    return answer;
}
